### ISSUE TITLE
Actual cumulative weight implementation doesn't match as document describe?

### ISSUE THREAD

**Issue by wusyong** [iotaledger/iri#847]
According [tip sel. documentation](https://hackmd.io/o1dVVX8lSqejkulnXok6-g#22-Cumulative-Weight-Implementation), set of all direct/indirect approvers are called future set. But in this function `sortTransactionsInTopologicalOrder` doesn't really record all txs. In other words, the size of future set doesn't seem right. Doesn't it seem there should be a for loop for the following section?

https://github.com/iotaledger/iri/blob/dev/src/main/java/com/iota/iri/service/tipselection/impl/CumulativeWeightCalculator.java#L43
```java
Collection<Hash> appHashes = getTxDirectApproversHashes(txHash, txToDirectApprovers);
                if (CollectionUtils.isNotEmpty(appHashes)) {
                    Hash txApp = getAndRemoveApprover(appHashes);
                    stack.push(txApp);
                    continue;
```

**Comment by GalRogozinski** [comment]
The purpose of the method is to perform topological sort on a subtangle. From the entry point it scans all reachable transactions and outputs them in a certain order. There are unit tests that ascertain the method is written correctly. This one for example ,https://github.com/iotaledger/iri/blob/dev/src/test/java/com/iota/iri/service/tipselection/impl/CumulativeWeightCalculatorTest.java#L198-L227, compares the result of the new efficient implementation with a naive less efficient implementation.

Note that topological order is not deterministic by definition, this is why the unit tests check for the cumulative weight which is deterministic.

In short, I didn't understand why you think the method is broken. Maybe write a unit test that proves your claim?

**Comment by wusyong** [comment]
I'm not saying the method is broken. It just seems that `appHashes` doesn't store all tx hash for me. But yes looks like it's better to test before my assumption. I will look into it in the future.
